### ISSUE TITLE
BUG: using msvc + meson to build scipy --> cl cannot be used to compile programs

### ISSUE THREAD

**Issue by andyfaff** [scipy/scipy#16935]
### Describe your issue.

I'm investigating building on Windows on a local machine. Rather than using mingw-w64 from rtools I thought I'd start to try using a conda based environment, to see it's possible to setup an environment that would successfully build scipy.

I setup the environment and try to compile, but I get an issue with the (MSVC) C compiler straight away. It seems that a flag is being added, `/Wno-unused-function` that cl doesn't know about. Is this something that scipy could fix, or is this an upstream meson thing? I've tried building a simpler package and there were no issues:

```
:: this simple test works
mkdir meson-test
cd meson-test
meson init
meson build
```

(I'm also looking into using the `clang-cl` compiler supplied by the `compilers` conda-forge package, but can't get as far as building the simple test above).

### Reproducing Code Example

```python
conda create -n test
conda activate test
conda install -c conda-forge compilers python=3.10 meson

meson build
```

### Error message

```shell
Build started at 2022-08-31T10:31:58.903013
Main binary: C:\Users\anz\Miniconda3\envs\test\python.exe
Build Options: 
Python system: Windows
The Meson build system
Version: 0.63.1
Source dir: \\filer\home\anz\programming\scipy
Build dir: \\filer\home\anz\programming\scipy\build
Build type: native build
Project name: SciPy
Project version: 1.10.0.dev0
-----
Detecting compiler via: icl ""
-----
Detecting compiler via: cl /?
compiler returned <Popen: returncode: 0 args: ['cl', '/?']>
compiler stdout:
usage: cl [ option... ] filename... [ /link linkoption... ]

compiler stderr:
Microsoft (R) C/C++ Optimizing Compiler Version 19.16.27048 for x64
Copyright (C) Microsoft Corporation.  All rights reserved.


Sanity testing C compiler: cl
Is cross compiler: False.
Sanity check compiler command line: cl sanitycheckc.c /Fesanitycheckc.exe /MD -Wno-unused-function -Wno-conversion -Wno-misleading-indentation -Wno-incompatible-pointer-types /nologo /showIncludes /utf-8 /link
Sanity check compile stdout:

-----
Sanity check compile stderr:
cl : Command line error D8021 : invalid numeric argument '/Wno-unused-function'

-----

\\filer\home\anz\programming\scipy\meson.build:1:0: ERROR: Compiler cl can not compile programs.
```

### SciPy/NumPy/Python version information

Python 3.10

**Comment by eli-schwartz** [comment]
> It seems that a flag is being added, `/Wno-unused-function` that cl doesn't know about. Is this something that scipy could fix, or is this an upstream meson thing? I've tried building a simpler package and there were no issues:

Grepping the SciPy repo for `Wno-unused-function` shows a bunch of hits, all being unconditionally added to e.g. `cpp_args:` on various modules, as well as set in `default_options: ['c_args=-Wno-unused-function -Wno-conversion -Wno-misleading-indentation -Wno-incompatible-pointer-types']` for the entire repository.

Aside: I'd probably use add_project_arguments for this, since I might want to do, say, `-Dc_args="-march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fexceptions -Wp,-D_FORTIFY_SOURCE=2 -Wformat -Werror=format-security -fstack-clash-protection -fcf-protection"` for my standard OS hardening and CPU tuning flags. And that would wipe the defaults.

More generally, Meson has `cxx.get_supported_arguments()` to check a bunch of optional flags, and only pass the ones the compiler actually supports. That might be a good fit here, because they aren't mission-critical flags.

**Comment by rgommers** [comment]
MSVC (or `clang-cl`) + Intel Fortran is a valid combination to build with, it just hasn't been tested yet. I think this is the first bug report for it. Unfortunately compiler warning flags aren't very portable. What @eli-schwartz suggests here should fix it.

**Comment by rgommers** [comment]
I don't think you can get Intel Fortran with `conda` by the way, only the runtime. So you probably have to install it from https://www.intel.com/content/www/us/en/developer/tools/oneapi/hpc-toolkit-download.html?operatingsystem=window&distributions=webdownload&options=online

**Comment by ilayn** [comment]
I tried these options before but never managed to make it working (both ifort and gfortran). If these flags are not used or critical, then let's get rid of them. Because we are not adding C++ code that often and we can turn them on in PRs only.

**Comment by rgommers** [comment]
Ideally we get rid of all those global args by the way - each one identifies a problem, mostly in `f2py`-generated code. I'm just finding the last issues with `-Wno-incompatible-pointer-types`. It's a little painful though:(

> If these flags are not used or critical, then let's get rid of them

They are necessary to make GCC build with `-Werror` in a CI job, which is critical. So adding them conditionally is the way to go, that should be 100% safe.

**Comment by rgommers** [comment]
I'll have a go at a fix here.

**Comment by andyfaff** [comment]
In my naivety Im trying to build scipy with compilers only available from conda. I'm currently intending to try flang as the Fortran compiler

**Comment by rgommers** [comment]
Ah, curious to see how far that gets you today. There are still issues with Flang last time we looked at it, see gh-15290.

**Comment by andyfaff** [comment]
I think the flang on conda-forge is *classic flang*. It was busy building OpenBLAS when I left it last night.

**Comment by andyfaff** [comment]
I managed to build OpenBLAS with flang and installed into `c:\opt`. After altering `site.cfg` I tried `python setup.py bdist_wheel`. I can get the wheel to build (using clang-cl+flang). I then used `delvewheel` to repair the wheel.

When I run the tests via `python -c "import scipy;scipy.test()"` there are various errors along the lines of `ImportError: DLL load failed while importing _iterative: The specified module could not be found.`. If I navigate to a directory containing all the shared libs that are bundled into the wheel and retry the test command then those errors disappear (indicating that the delvewheel repair didn't work fully). The test suite runs up to a point then segfaults.
